### PR TITLE
Deprecate write_predictions on the LightningModule

### LINKED ISSUE #7065: Deprecate LightningModule's write_predictions in favor of predict_step
## 🚀 Feature
Not a feature but a request for eventual removal: 
The LightningModule has `write_predictions` and `write_prediction_dict` as an API. The LightningModule also now has [predict_step](https://github.com/PyTorchLightning/pytorch-lightning/blob/7b0b0d284494d08e3983321d0cc42fe9e5faeb41/pytorch_lightning/core/lightning.py#L1074-L1078
), there's `Trainer.predict()`, and users can configure callbacks for this loop based on the predict hooks.

This means we can replace `write_predictions` with `predict_step` + a callback to output results to any destination (not just a file), which is more flexible than `write_predictions`

### Motivation

Avoid confusion/redundancy for users as to which hook they should implement

`write_predictions` also reaches deep into the trainer internals which is bad practice. this also implies that the trainer is available, which is not strictly necessary for pure inference: as a user I should be able to generate batches on my own and run the forward pass without requiring access to the trainer or its internal evaluation loop

### Pitch

Mark these APIs on the Lightning Module as deprecated in v1.3 and slated fo removal in v1.5. Replace `write_predictions` with `predict_step` + a callback to output results to any destination (not just a file), which is more flexible than `write_predictions`

### Alternatives

Keep as is

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #7065

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -405,7 +405,14 @@ def write_prediction(
             when running in distributed mode, calling ``write_prediction`` will create a file for
             each device with respective names: ``filename_rank_0.pt``, ``filename_rank_1.pt``, ...
 
+        .. deprecated::v1.3
+            Will be removed in v1.5.0.
         """
+        rank_zero_deprecation(
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:411]: you can also use a deprecation wrapper as we have in metrics...
https://borda.github.io/pyDeprecate/

**Review by Borda — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Comment by tchaton** [comment]
Hey @ananthsub,

I was talking with @carmocca and we thought we might want to introduce a 
Writer Callback to Lightning for users to save predictions on ``batch / epoch``. It should work for all stages possibly and users could active them if needed.

We should add properly `on_predict_step_end` and `on_predict_epoch_end`  hooks.

We also noticed some issues with `trainer.predict(...)` with spawn (ddp_spawn, tpu_spawn). Only rank 0 results are being returned to the main process. This would be an issue if those accelerators are used to serve models into `endpoints`.

Should we try to return all ranks results for predict ?

Best,
T.C

**Comment by ananthsub** [comment]
@tchaton I think it'd make sense to write independently from each rank instead of gathering all results on rank 0. In terms of returning results to users, couldn't the callback contain the return state, similar to how the model checkpoint callback tracks the top K model paths?

**Comment by awaelchli** [comment]
What do we recommend instead of write_predictions? The predict method of Trainer is only for when the model is already trained and needs a dataloader.

**Comment by carmocca** [comment]
@awaelchli The alternative is #7127

**Comment by awaelchli** [comment]
I have seen that but is it a replacement? write_predictions was a method that users could use anywhere in their LightningModule (whether we recommend that or not). I have not seen an option for this in #7127.

**Comment by ananthsub** [comment]
@awaelchli users can still use `torch.save` + file I/O from within the lightning module to save arbitrary values. I don't think there has to be a direct lightning API replacement in order to deprecate this

**Review by tchaton — APPROVED** [review]
LGTM !
